packages/library-base/src/commonMain/kotlin/io/realm/query/RealmQuery.kt

```diff
@@ -0,0 +1,371 @@
+/*
+ * Copyright 2021 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.query
+
+import io.realm.MutableRealm
+import io.realm.Realm
+import io.realm.RealmConfiguration
+import io.realm.RealmInstant
+import io.realm.RealmList
+import io.realm.RealmObject
+import io.realm.RealmResults
+import io.realm.internal.Flowable
+import kotlinx.coroutines.flow.Flow
+import kotlin.reflect.KClass
+
+/**
+ * A `RealmQuery` encapsulates a query on a [Realm], a [RealmResults] or a [RealmList] instance
+ * using the `Builder` pattern. The query is executed using either [find] or subscribing to the
+ * [Flow] returned by [asFlow].
+ *
+ * If a [RealmObject] class is refactored, care has to be taken not to break any query filters.
+ *
+ * A [Realm] is unordered, which means that there is no guarantee that a query will return the
+ * objects in the order they where inserted. Use the [sort] functions if a specific order is
+ * required.
+ *
+ * Results are obtained quickly most of the times when using [find]. However, launching heavy
+ * queries from the UI thread may result in a drop of frames or even ANRs. If you want to prevent
+ * these behaviors, you can use [asFlow] and collect the results asynchronously.
+ *
+ * @param E the class of the objects to be queried.
+ */
+interface RealmQuery<E> : RealmElementQuery<E> {
```

Comment:
After our discission, Isn't this always going to be a RealmObject?

Reply:
See other related discussions.